barretenberg/cpp/src/barretenberg/dsl/acir_proofs/acir_composer.cpp

```diff
@@ -81,17 +81,26 @@ std::vector<uint8_t> AcirComposer::create_proof(acir_format::acir_format& constr
 void AcirComposer::create_goblin_circuit(acir_format::acir_format& constraint_system,
                                          acir_format::WitnessVector& witness)
 {
-    // Provide the builder with the op queue owned by the goblin instance
-    goblin_builder_.op_queue = goblin.op_queue;
-
-    create_circuit_with_witness(goblin_builder_, constraint_system, witness);
+    // The public inputs in constraint_system do not index into "witness" but rather into the future "variables" which
+    // it assumes will be equal to witness but with a prepended zero. We want to remove this +1 so that public_inputs
+    // properly indexes into witness bc we're about to make calls like add_variable(witness[public_inputs[idx]]). Once
+    // the +1 is removed from noir, this correction can be removed entirely and we can use
+    // constraint_system.public_inputs directly.
+    const uint32_t pre_applied_noir_offset = 1;
+    std::vector<uint32_t> corrected_public_inputs;
+    for (const auto& index : constraint_system.public_inputs) {
+        corrected_public_inputs.emplace_back(index - pre_applied_noir_offset);
+    }
 
-    info("after create_circuit_with_witness: num_gates = ", goblin_builder_.num_gates);
+    // Construct a builder using the witness and public input data from acir
+    goblin_builder_ =
+        acir_format::GoblinBuilder{ goblin.op_queue, witness, corrected_public_inputs, constraint_system.varnum };
 
-    // Correct for the addition of const variables in the builder constructor
-    acir_format::apply_wire_index_offset(goblin_builder_);
+    // Populate constraints in the builder via the data in constraint_system
+    acir_format::build_constraints(goblin_builder_, constraint_system, true);
 
-    // Add some arbitrary op gates to ensure the associated polynomials are non-zero
+    // HACK: Add some arbitrary op gates to ensure the associated polynomials are non-zero and to give ECCVM and
+    // Translator some ECC ops to process.
     GoblinTestingUtils::construct_goblin_ecc_op_circuit(goblin_builder_);
```

Comment:
Is there an issue open for this? Its not blocking this PR as it was already there before this PR was merged.

Would be good to clean this up in a near-future PR as having TestingUtils being used anywhere except for tests seems like a footgun

Reply:
Added the relevant issue. The hope is to get all of this sorted out in the near future